controller/src/controller.rs

```diff
@@ -271,27 +272,60 @@ impl<T: BottlerocketShadowClient> BrupopController<T> {
+            // fixed times: 0 10 * * * */At 10:00 on Monday.
+            // periods of time (time window): 0 10-12 * * mon */At every minute on every Monday from 10 through 12
+
+            // brupop controller needs to use different logics to deal with fixed times or periods of time (time window)
+            // => fixed times: trigger brupop update and complete all waitingForUpdate nodes.
+            // => periods of time (time window): trigger brupop update within time window. If current time isn't within the time window,
             // controller shouldn't have any action on it.
-            match within_time_window()? {
-                // If controller has already started a node update but the update time window stops,
-                // controller should respect it, continue to complete the update, and stop actions on remaining nodes.
-                // This logic will cooperate with 6 mins buffer strategy.
-                false => {
-                    // try to find out if any nodes are being updated. If yes, controller
-                    // will complete them and pause actions on other waitingForUpdate nodes.
-                    if !active_set.is_empty() {
-                        self.progress_active_set(active_set).await?;
-                    } else {
-                        sleep(ACTION_INTERVAL).await;
-                        continue;
-                    }
-                }
+            match is_time_window(brupop_schedule.clone())? {
                 true => {
-                    if !active_set.is_empty() {
-                        self.progress_active_set(active_set).await?;
+                    // If current time isn't within the time window, controller shouldn't have any action on it and sleep until next scheduled time point.
+                    match within_time_window(brupop_schedule.clone()) {
+                        true => {
+                            // Bring one more node each time if the active nodes size is less than MAX_CONCURRENT_UPDATE setting.
+                            if active_set_size < get_max_concurrent_update()? {
+                                // If there's nothing to operate on, check to see if any other nodes are ready for action.
+                                let new_active_node = self.find_and_update_ready_brs().await?;
+                                if let Some(brs) = new_active_node {
+                                    event!(Level::INFO, name = %brs.name_any(), "Began updating new node.")
+                                }
+                            }
+                        }
+                        false => {
+                            event!(
+                                Level::INFO,
+                                "stop finding ready brs and complete remaining actions."
+                            );
+                        }
                     }
-
+                }
+                false => {
```

Comment:
As designed, when users set fixed times, brupop will update all waitingForUpdate nodes. 

>It seems like if we're doing fixed times, we only update MAX_CONCURRENT_UPDATE nodes before sleeping til the next time window.

Actually it will update all nodes not MAX_CONCURRENT_UPDATE nodes. When controller detects the cron expression is fixed time, it will keep updating the nodes on  `active_set` and once the size of `active_set` is beyond  `MAX_CONCURRENT_UPDATE `, and it will find more ready nodes and add to `active_set`. Thus, this logic will support controller to update all nodes on the cluster. This logic is kind of tricky, and I'd like to have  a call if you still have question. thanks!
```
loop{
    let active_set = self.active_brs_set();
    let active_set_size = active_set.len();
    event!(Level::TRACE, ?active_set, "Found active set of nodes.");

    // complete any ongoing updates even if we are outside of a scheduled window.
    if !active_set.is_empty() {
        self.progress_active_set(active_set).await?;
    } else {
     ...........
     }
    if is_time_window(&brupop_schedule)? {
    }  else {
    // Bring one more node each time if the active nodes size is less than MAX_CONCURRENT_UPDATE setting.
    if active_set_size < get_max_concurrent_update()? {
        // If there's nothing to operate on, check to see if any other nodes are ready for action.
        let new_active_node = self.find_and_update_ready_brs().await?;
        if let Some(brs) = new_active_node {
            event!(Level::INFO, name = %brs.name_any(), "Began updating new node.")
        }
    }
  }
}
```